File: src/tables/kern.rs
use std::collections::HashSet;

use crate::parser::*;
use crate::{TitleKind, ValueType, Error, Result};

pub fn parse(parser: &mut Parser) -> Result<()> {
    // The `kern` table has two variants: OpenType one and Apple one.
    // And they both have different headers.
    // There are no robust way to distinguish them, so we have to guess.
    //
    // The OpenType one has the first two bytes (UInt16) as a version set to 0.
    // While Apple one has the first four bytes (Fixed) set to 1.0
    // So the first two bytes in case of an OpenType format will be 0x0000
    // and 0x0001 in case of an Apple format.
    let version = parser.peek::<u16>()?;
    if version == 0 {
        ot::parse(parser)?;
    } else {
        aat::parse(parser)?;
    }

    Ok(())
}

// https://docs.microsoft.com/en-us/typography/opentype/spec/kern
mod ot {
    use super::*;

    #[derive(Clone, Copy, Debug)]
    struct Coverage(u8);

    impl FromData for Coverage {
        const NAME: ValueType = ValueType::BitFlags;

        #[inline]
        fn parse(data: &[u8]) -> Result<Self> {
            Ok(Coverage(data[0]))
        }
    }

    impl std::fmt::Display for Coverage {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            use std::fmt::Write;
            let mut s = String::new();

            let bits = U8Bits(self.0);
            writeln!(&mut s, "{}", bits)?;
            if bits[0] { s.push_str("Bit 0: Horizontal\n"); }
            if bits[1] { s.push_str("Bit 1: Has minimum values\n"); }
            if bits[2] { s.push_str("Bit 2: Cross-stream\n"); }
            if bits[3] { s.push_str("Bit 3: Override\n"); }
            // 4-7 - reserved

            s.pop(); // pop last new line

            f.write_str(&s)
        }
    }

    pub fn parse(parser: &mut Parser) -> Result<()> {
        parser.read::<u16>("Version")?;
        let number_of_tables = parser.read::<u16>("Number of tables")?;
        for _ in 0..number_of_tables {
            let subtable_start = parser.offset();

            parser.begin_group("");
            parser.read::<u16>("Version")?;
            parser.read::<u16>("Length")? as usize;
            let format = parser.read::<u8>("Format")?;
            parser.read::<Coverage>("Coverage")?;

            match format {
                0 => aat::parse_format0(parser)?,
                2 => aat::parse_format2(subtable_start, parser)?,
                _ => return Err(Error::Custom(format!("{} is not a valid format", format))),
            }

            parser.end_group_with_title(format!("Table Format {}", format));
        }

        Ok(())
    }
}

// https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6kern.html
mod aat {
    use super::*;

    #[derive(Clone, Copy, Debug)]
    struct Coverage(u8);

    impl FromData for Coverage {
        const NAME: ValueType = ValueType::BitFlags;

        #[inline]
        fn parse(data: &[u8]) -> Result<Self> {
            Ok(Coverage(data[0]))
        }
    }

    impl std::fmt::Display for Coverage {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            use std::fmt::Write;
            let mut s = String::new();

            let bits = U8Bits(self.0);
            writeln!(&mut s, "{}", bits)?;
            // 0-4 - reserved
            if bits[5] { s.push_str("Bit 5: Has variation\n"); }
            if bits[6] { s.push_str("Bit 6: Cross-stream\n"); }
            if bits[7] { s.push_str("Bit 7: Vertical\n"); }

            s.pop(); // pop last new line

            f.write_str(&s)
        }
    }


    #[derive(Clone, Copy, Debug)]
    struct EntryFlags(u16);

    impl FromData for EntryFlags {
        const NAME: ValueType = ValueType::BitFlags;

        #[inline]
        fn parse(data: &[u8]) -> Result<Self> {
            u16::parse(data).map(EntryFlags)
        }
    }

    impl std::fmt::Display for EntryFlags {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            use std::fmt::Write;
            let mut s = String::new();

            let bits = U16Bits(self.0);
            writeln!(&mut s, "{}", self.0 & 0x3FFF)?;
            writeln!(&mut s, "Offset {}", bits)?;
            if bits[15] { s.push_str("Bit 15: Push onto the kerning stack\n"); }

            s.pop(); // pop last new line

            f.write_str(&s)
        }
    }


    #[derive(Clone, Copy, Debug)]
    struct Action(u16);

    impl FromData for Action {
        const NAME: ValueType = ValueType::BitFlags;

        #[inline]
        fn parse(data: &[u8]) -> Result<Self> {
            u16::parse(data).map(Action)
        }
    }

    impl std::fmt::Display for Action {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            match self.0 {
                0x0001 => f.write_str("Kerning 0. End of List."),
                0x8001 => f.write_str("Reset cross-stream. End of List."),
                _ => write!(f, "Kerning {}", self.0 as i16),
            }
        }
    }


    struct Entry {
        new_state: u16,
        #[allow(dead_code)]
        flags: u16,
    }


    pub fn parse(parser: &mut Parser) -> Result<()> {
        parser.read::<Fixed>("Version")?;
        let number_of_tables = parser.read::<u32>("Number of tables")?;
        for _ in 0..number_of_tables {
            let subtable_start = parser.offset();

            parser.begin_group("");
            let length = parser.read::<u32>("Length")? as usize;
            // Yes, the coverage and format order is inverted in AAT.
            parser.read::<Coverage>("Coverage")?;
            let format = parser.read::<u8>("Format")?;
            parser.read::<u16>("Tuple index")?;

            match format {
                0 => parse_format0(parser)?,
                1 => parse_format1(length, parser)?,
                2 => parse_format2(subtable_start, parser)?,
                3 => parse_format3(subtable_start, length, parser)?,
                _ => return Err(Error::Custom(format!("{} is not a valid format", format))),
            }

            parser.end_group_with_title(format!("Table Format {}", format));
        }

        Ok(())
    }

    pub fn parse_format0(parser: &mut Parser) -> Result<()> {
        let count = parser.read::<u16>("Number of kerning pairs")?;
        parser.read::<u16>("Search range")?;
        parser.read::<u16>("Entry selector")?;
        parser.read::<u16>("Range shift")?;

        parser.begin_group("Values");
        for _ in 0..count {
            parser.begin_group("Pair");
            parser.read::<GlyphId>("Left")?;
            parser.read::<GlyphId>("Right")?;
            parser.read::<i16>("Value")?;
            parser.end_group();
        }
        parser.end_group();

        Ok(())
    }

    fn parse_format1(subtable_length: usize, parser: &mut Parser) -> Result<()> {
        // Sadly, the *State Table for Contextual Kerning* format
        // isn't really documented.
        // The AAT kern documentation has only a rough outline and a single example,
        // which is clearly not enough to understand how it actually stored.
        // And the only FOSS application I know that supports it is HarfBuzz,
        // so we are borrowing its implementation.
        // No idea how long it took them to figure it out.
        //
        // Also, there is an obsolete tool called `ftxanalyzer`,
        // from the *Apple Font Tool Suite* that can dump an AAT kern table as an XML.
        // But the result is still pretty abstract and it's hard to understand
        // how it was produced.
        // It's not trivial to install it on a modern Mac, so here is a useful
        // instruction (not mine):
        // https://gist.github.com/thetrevorharmon/9afdeb41a74f8f32b9561eeb83b10eff
        //
        // And here are some Apple fonts that actually use this format
        // (at least in macOS 10.14):
        // - Diwan Thuluth.ttf
        // - Farisi.ttf
        // - GeezaPro.ttc
        // - Mishafi.ttf
        // - Waseem.ttc

        let start = parser.offset();

        let num_of_classes = parser.read::<u16>("Number of classes")?;
        // Note that offsets are not from the subtable start,
        // but from subtable start + header.
        let class_table_offset = parser.read::<Offset16>("Offset to class subtable")?.to_usize();
        let state_array_offset = parser.read::<Offset16>("Offset to state array")?.to_usize();
        let entry_table_offset = parser.read::<Offset16>("Offset to entry table")?.to_usize();
        let values_offset = parser.read::<Offset16>("Offset to values")?.to_usize();

        // Check that offsets are properly ordered.
        // We do not support random order.
        assert!(class_table_offset < state_array_offset);
        assert!(state_array_offset < entry_table_offset);
        assert!(entry_table_offset < values_offset);

        let number_of_entries = {
            // Collect states.
            parser.jump_to(start + state_array_offset)?;
            let mut simple = parser.to_simple();
            // We don't know the length of the states yet,
            // so use all the data from offset to the end of the subtable.
            let states = simple.read_bytes(subtable_length - state_array_offset)?.to_vec();

            // Collect entries.
            parser.jump_to(start + entry_table_offset)?;
            let mut simple = parser.to_simple();
            // We don't know the length of the states yet,
            // so use all the data from offset to the end of the subtable.
            let entries_count = (subtable_length - entry_table_offset) / 4;
            let mut entries = Vec::new();
            for _ in 0..entries_count {
                entries.push(Entry {
                    new_state: simple.read()?,
                    flags: simple.read()?,
                });
            }

            detect_number_of_entries(
                num_of_classes, state_array_offset as u16, &states, &entries,
            )?
        };

        parser.jump_to(start + class_table_offset)?;
        parser.begin_group("Class subtable");
        parser.read::<GlyphId>("First glyph")?;
        let num_glyphs = parser.read::<u16>("Number of glyphs")?;
        parser.read_array::<u8>("Classes", TitleKind::Class, num_glyphs as usize)?;
        parser.end_group();

        parser.jump_to(start + state_array_offset)?;
        parser.begin_group("State array");
        // We only assume that an *entry table* is right after *state array*.
        for _ in 0..((entry_table_offset - state_array_offset) / num_of_classes as usize) {
            parser.read_bytes(num_of_classes as usize, "State")?;
        }
        parser.end_group();

        parser.jump_to(start + entry_table_offset)?;
        parser.begin_group_with_value("Entry table", number_of_entries.to_string());
        for i in 0..number_of_entries {
            parser.begin_group(format!("Entry {}", i));
            parser.read::<Offset16>("State offset")?;
            parser.read::<EntryFlags>("Flags")?;
            parser.end_group();
        }
        parser.end_group();

        parser.jump_to(start + values_offset)?;
        let num_of_actions = ((subtable_length - 8) - (parser.offset() - start)) / 2;
        parser.read_array::<Action>("Actions", TitleKind::Action, num_of_actions)?;

        Ok(())
    }

    // Based on: https://github.com/harfbuzz/harfbuzz/blob/5b91c52083aee1653c0cf1e778923de00c08fa5d/src/hb-aat-layout-common.hh#L524
    fn detect_number_of_entries(
        num_of_classes: u16,
        state_array_offset: u16,
        states: &[u8],
        entries: &[Entry],
    ) -> Result<u32> {
        let num_of_classes = num_of_classes as i32;

        let mut min_state: i32 = 0;
        let mut max_state: i32 = 0;
        let mut num_entries: u32 = 0;

        let mut state_pos: i32 = 0;
        let mut state_neg: i32 = 0;
        let mut entry: u32 = 0;
        let mut max_ops: i32 = 0x3FFFFFFF;
        while min_state < state_neg || state_pos <= max_state {
            if min_state < state_neg {
                // Negative states.

                max_ops -= state_neg - min_state;
                if max_ops <= 0 {
                    return Err(Error::Custom("invalid state machine".to_string()));
                }

                // Sweep new states.
                let end = min_state * num_of_classes;
                for i in (0..end).rev() {
                    num_entries = std::cmp::max(num_entries, states[(i-1) as usize] as u32 + 1);
                }

                state_neg = min_state;
            }

            if state_pos <= max_state {
                // Positive states.

                max_ops -= max_state - state_pos + 1;
                if max_ops <= 0 {
                    return Err(Error::Custom("invalid state machine".to_string()));
                }

                // Sweep new states.
                let start = state_pos * num_of_classes;
                let end = (max_state + 1) * num_of_classes;
                for i in start..end {
                    num_entries = std::cmp::max(num_entries, states[i as usize] as u32 + 1);
                }

                state_pos = max_state + 1;
            }

            max_ops -= (num_entries - entry) as i32;
            if max_ops <= 0 {
                return Err(Error::Custom("invalid state machine".to_string()));
            }

            // Sweep new entries.
            for i in entry..num_entries {
                let new_state = (entries[i as usize].new_state as i32 - state_array_offset as i32)
                    / num_of_classes as i32;

                min_state = std::cmp::min(min_state, new_state);
                max_state = std::cmp::max(max_state, new_state);
            }

            entry = num_entries;
        }

        Ok(num_entries)
    }

    pub fn parse_format2(subtable_start: usize, parser: &mut Parser) -> Result<()> {
        // Apple fonts that are using this table: Apple Chancery.ttf

        parser.read::<u16>("Row width in bytes")?;

        enum OffsetKind {
            LeftHandTable,
            RightHandTable,
            Array,
        }

        let left_hand_table_offset = parser.read::<Offset16>("Offset to left-hand class table")?.to_usize();
        let right_hand_table_offset = parser.read::<Offset16>("Offset to right-hand class table")?.to_usize();
        let array_offset = parser.read::<Offset16>("Offset to kerning array")?.to_usize();

        let rows = detect_number_of_classes(subtable_start + left_hand_table_offset, parser)?;
        let columns = detect_number_of_classes(subtable_start + right_hand_table_offset, parser)?;

        // We have to sort offsets, because they can be in any order.
        let mut offsets = [
            (OffsetKind::LeftHandTable, left_hand_table_offset),
            (OffsetKind::RightHandTable, right_hand_table_offset),
            (OffsetKind::Array, array_offset),
        ];
        offsets.sort_by_key(|v| v.1);

        for (kind, offset) in offsets.iter() {
            match kind {
                OffsetKind::LeftHandTable => {
                    parser.jump_to(subtable_start + offset)?;
                    parser.begin_group("Left-hand class table");
                    parser.read::<GlyphId>("First glyph")?;
                    let count = parser.read::<u16>("Number of glyphs")?;
                    parser.read_array::<u16>("Classes", TitleKind::Class, count as usize)?;
                    parser.end_group();
                }
                OffsetKind::RightHandTable => {
                    parser.jump_to(subtable_start + offset)?;
                    parser.begin_group("Right-hand class table");
                    parser.read::<GlyphId>("First glyph")?;
                    let count = parser.read::<u16>("Number of glyphs")?;
                    parser.read_array::<u16>("Classes", TitleKind::Class, count as usize)?;
                    parser.end_group();
                }
                OffsetKind::Array => {
                    parser.jump_to(subtable_start + offset)?;
                    parser.read_array::<i16>("Kerning values", TitleKind::Value, rows * columns)?;
                }
            }
        }

        Ok(())
    }

    fn detect_number_of_classes(offset: usize, parser: &mut Parser) -> Result<usize> {
        parser.jump_to(offset)?;
        let mut parser = parser.to_simple();
        parser.read::<GlyphId>()?;
        let count = parser.read::<u16>()?;
        let mut classes = HashSet::new();
        for _ in 0..count {
            classes.insert(parser.read::<u16>()?);
        }

        Ok(classes.len())
    }

    fn parse_format3(subtable_start: usize, subtable_len: usize, parser: &mut Parser) -> Result<()> {
        // Apple fonts that are using this table: Skia.ttf

        let glyph_count = parser.read::<u16>("Number of glyphs")?;
        let kern_values = parser.read::<u8>("Number of kerning values")?;
        let left_hand_classes = parser.read::<u8>("Number of left-hand classes")?;
        let right_hand_classes = parser.read::<u8>("Number of right-hand classes")?;
        parser.read::<u8>("Reserved")?;

        parser.read_array::<i16>("Kerning values", TitleKind::Value, kern_values as usize)?;
        parser.read_array::<u8>("Left-hand classes", TitleKind::Class, glyph_count as usize)?;
        parser.read_array::<u8>("Right-hand classes", TitleKind::Class, glyph_count as usize)?;
        parser.read_array::<u8>("Indices", TitleKind::Index,
                                left_hand_classes as usize * right_hand_classes as usize)?;

        let left = subtable_len as isize - (parser.offset() - subtable_start) as isize;
        if left > 0 {
            parser.read_bytes(left as usize, "Padding")?;
        }

        Ok(())
    }
}

File: src/tables/kern.cpp
#include <bitset>

#include "src/algo.h"
#include "tables.h"

struct OpenTypeCoverage
{
    static const int Size = 1;
    static const QString Type;

    static OpenTypeCoverage parse(const quint8 *data)
    { return { data[0] }; }

    static QString toString(const OpenTypeCoverage &value)
    {
        std::bitset<8> bits(value.d);
        auto flagsStr = QString::fromUtf8(bits.to_string().c_str()) + '\n';

        if (bits[0]) { flagsStr += "Bit 0: Horizontal\n"; }
        if (bits[1]) { flagsStr += "Bit 1: Has minimum values\n"; }
        if (bits[2]) { flagsStr += "Bit 2: Cross-stream\n"; }
        if (bits[3]) { flagsStr += "Bit 3: Override\n"; }
        // 4-7 - reserved

        flagsStr.chop(1); // trim trailing newline

        return flagsStr;
    }

    quint8 d;
};

const QString OpenTypeCoverage::Type = Parser::BitflagsType;


struct AppleCoverage
{
    static const int Size = 1;
    static const QString Type;

    static AppleCoverage parse(const quint8 *data)
    { return { data[0] }; }

    static QString toString(const AppleCoverage &value)
    {
        std::bitset<8> bits(value.d);
        auto flagsStr = QString::fromUtf8(bits.to_string().c_str()) + '\n';

        // 0-4 - reserved
        if (bits[5]) { flagsStr += "Bit 5: Has variation\n"; }
        if (bits[6]) { flagsStr += "Bit 6: Cross-stream\n"; }
        if (bits[7]) { flagsStr += "Bit 7: Vertical\n"; }

        flagsStr.chop(1); // trim trailing newline

        return flagsStr;
    }

    quint8 d;
};

const QString AppleCoverage::Type = Parser::BitflagsType;


static void parseFormat0(Parser &parser)
{
    const auto count = parser.read<UInt16>("Number of kerning pairs");
    parser.read<UInt16>("Search range");
    parser.read<UInt16>("Entry selector");
    parser.read<UInt16>("Range shift");

    parser.readArray("Values", count, [&](const auto index){
        parser.beginGroup(index);
        parser.read<GlyphId>("Left");
        parser.read<GlyphId>("Right");
        parser.read<Int16>("Value");
        parser.endGroup();
    });
}

namespace Format1 {
    struct EntryFlags
    {
        static const int Size = 2;
        static const QString Type;

        static EntryFlags parse(const quint8 *data)
        { return { UInt16::parse(data) }; }

        static QString toString(const EntryFlags &value)
        {
            std::bitset<16> bits(value.d);
            auto flagsStr = "Offset " + numberToString(value.d & 0x3FFF) + '\n';
            flagsStr += QString::fromUtf8(bits.to_string().c_str()) + '\n';

            if (bits[15]) { flagsStr += "Bit 15: Push onto the kerning stack\n"; }

            flagsStr.chop(1); // trim trailing newline

            return flagsStr;
        }

        quint16 d;
    };

    const QString EntryFlags::Type = Parser::BitflagsType;


    struct Action
    {
        static const int Size = 2;
        static const QString Type;

        static Action parse(const quint8 *data)
        { return { UInt16::parse(data) }; }

        static QString toString(const Action &value)
        {
            static const QString value1 = QLatin1String("Kerning 0. End of List.");
            static const QString value2 = QLatin1String("Reset cross-stream. End of List.");

            if (value.d == 0x0001) {
                return value1;
            } else if (value.d == 0x8001) {
                return value1;
            } else {
                return QString("Kerning %1").arg(qint16(value.d));
            }
        }

        quint16 d;
    };

    const QString Action::Type = QLatin1String("Action");


    struct Entry {
        quint16 newState;
        quint16 flags;
    };


    // Based on: https://github.com/harfbuzz/harfbuzz/blob/5b91c52083aee1653c0cf1e778923de00c08fa5d/src/hb-aat-layout-common.hh#L524
    static quint32 detectNumberOfEntries(
        const qint32 numberOfClasses,
        const quint16 stateArrayOffset,
        const QByteArray states,
        const QVector<Entry> &entries
    ) {
        qint32 minState = 0;
        qint32 maxState = 0;
        quint32 numEntries = 0;

        qint32 statePos = 0;
        qint32 stateNeg = 0;
        quint32 entry = 0;
        qint32 maxOps = 0x3FFFFFFF;
        while (minState < stateNeg || statePos <= maxState) {
            if (minState < stateNeg) {
                // Negative states.

                maxOps -= stateNeg - minState;
                if (maxOps <= 0) {
                    throw QString("invalid state machine");
                }

                // Sweep new states.
                const qint32 end = minState * numberOfClasses;
                if (end > 0) {
                    for (qint32 i = end-1; i >= 0; i--) {
                        if (i - 1 < states.size()) {
                            numEntries = qMax(numEntries, quint32(states[i - 1]) + 1);
                        } else {
                            throw QString("invalid index");
                        }
                    }
                }

                stateNeg = minState;
            }

            if (statePos <= maxState) {
                // Positive states.

                maxOps -= maxState - statePos + 1;
                if (maxOps <= 0) {
                    throw QString("invalid state machine");
                }

                // Sweep new states.
                const auto start = statePos * numberOfClasses;
                const auto end = (maxState + 1) * numberOfClasses;
                for (int i = start; i < end; i++) {
                    if (i < states.size()) {
                        numEntries = qMax(numEntries, quint32(states[i]) + 1);
                    } else {
                        throw QString("invalid index");
                    }
                }

                statePos = maxState + 1;
            }

            maxOps -= int(numEntries - entry);
            if (maxOps <= 0) {
                throw QString("invalid state machine");
            }

            // Sweep new entries.
            for (quint32 i = entry; i < numEntries; i++) {
                if (i >= quint32(entries.size())) {
                    throw QString("invalid index");
                }

                const auto newState = (int(entries[i].newState) - int(stateArrayOffset))
                    / int(numberOfClasses);

                minState = qMin(minState, newState);
                maxState = qMax(maxState, newState);
            }

            entry = numEntries;
        }

        return numEntries;
    }

    static void parse(const quint32 subtableSize, Parser &parser)
    {
        // Sadly, the *State Table for Contextual Kerning* format isn't really documented.
        // The AAT kern documentation has only a rough outline and a single example,
        // which is clearly not enough to understand how it actually stored.
        // And the only FOSS application I know that supports it is HarfBuzz,
        // so we are borrowing its implementation.
        // No idea how long it took them to figure it out.
        //
        // Also, there is an obsolete tool called `ftxanalyzer`,
        // from the *Apple Font Tool Suite* that can dump an AAT kern table as an XML.
        // But the result is still pretty abstract and it's hard to understand
        // how it was produced.
        // It's not trivial to install it on a modern Mac, so here is a useful
        // instruction (not mine):
        // https://gist.github.com/thetrevorharmon/9afdeb41a74f8f32b9561eeb83b10eff
        //
        // And here are some Apple fonts that actually use this format
        // (at least in macOS 10.14):
        // - Diwan Thuluth.ttf
        // - Farisi.ttf
        // - GeezaPro.ttc
        // - Mishafi.ttf
        // - Waseem.ttc

        const auto start = parser.offset();
        const auto shadow = parser.shadow();

        const quint16 numberOfClasses = parser.read<UInt16>("Number of classes");
        // Note that offsets are not from the subtable start,
        // but from subtable start + header.
        const auto classTableOffset = parser.read<Offset16>("Offset to class subtable");
        const auto stateArrayOffset = parser.read<Offset16>("Offset to state array");
        const auto entryTableOffset = parser.read<Offset16>("Offset to entry table");
        const auto valuesOffset = parser.read<Offset16>("Offset to values");

        // Check that offsets are properly ordered.
        // We do not support random order.
        Q_ASSERT(classTableOffset < stateArrayOffset);
        Q_ASSERT(stateArrayOffset < entryTableOffset);
        Q_ASSERT(entryTableOffset < valuesOffset);

        auto numberOfEntries = 0;
        {
            // Collect states.
            auto s1 = shadow;
            s1.advanceTo(stateArrayOffset);
            // We don't know the length of the states yet,
            // so use all the data from offset to the end of the subtable.
            const auto states = s1.readBytes(subtableSize - stateArrayOffset);

            // Collect entries.
            auto s2 = shadow;
            s2.advanceTo(entryTableOffset);
            // We don't know the length of the states yet,
            // so use all the data from offset to the end of the subtable.
            const auto entriesCount = (subtableSize - entryTableOffset) / 4;
            QVector<Entry> entries;
            for (quint32 i = 0; i < entriesCount; i++) {
                entries.append({
                    s2.read<UInt16>(),
                    s2.read<UInt16>(),
                });
            }

            numberOfEntries = detectNumberOfEntries(numberOfClasses, stateArrayOffset, states, entries);
        }

        parser.padTo(start + classTableOffset);
        parser.beginGroup("Class Subtable");
        parser.read<GlyphId>("First glyph");
        const auto numberOfGlyphs = parser.read<UInt16>("Number of glyphs");
        parser.readBasicArray<UInt8>("Classes", numberOfGlyphs);
        parser.endGroup();

        parser.padTo(start + stateArrayOffset);
        // We only assume that an *entry table* is right after *state array*.
        const auto arraysCount = (entryTableOffset - stateArrayOffset) / numberOfClasses;
        parser.readArray("State Array", arraysCount, [&](const auto /*index*/){
            parser.readBytes("Data", numberOfClasses);
        });

        parser.padTo(start + entryTableOffset);
        parser.readArray("Entries", numberOfEntries, [&](const auto index){
            parser.beginGroup(index);
            parser.read<Offset16>("State offset");
            parser.read<EntryFlags>("Flags");
            parser.endGroup();
        });

        parser.padTo(start + valuesOffset);
        const auto numberOfActions = ((subtableSize - 8) - (parser.offset() - start)) / 2;
        parser.readBasicArray<Action>("Actions", numberOfActions);
    }
}

static quint32 format2DetectNumberOfClasses(const quint32 offset, ShadowParser shadow) {
    shadow.advanceTo(offset);
    shadow.read<GlyphId>();
    const auto count = shadow.read<UInt16>();
    QSet<quint16> classes;
    for (quint32 i = 0; i < count; i++) {
        classes.insert(shadow.read<UInt16>());
    }

    return classes.size();
}

static void parseFormat2(const quint32 subtableStart, Parser &parser)
{
    // Apple fonts that are using this table: Apple Chancery.ttf

    const auto shadow = parser.shadow();
    const auto headerSize = parser.offset() - subtableStart;

    parser.read<UInt16>("Row width in bytes");

    const auto leftHandTableOffset = parser.read<Offset16>("Offset to left-hand class table");
    const auto rightHandTableOffset = parser.read<Offset16>("Offset to right-hand class table");
    const auto arrayOffset = parser.read<Offset16>("Offset to kerning array");

    const auto rows = format2DetectNumberOfClasses(leftHandTableOffset - headerSize, shadow);
    const auto columns = format2DetectNumberOfClasses(rightHandTableOffset - headerSize, shadow);

    enum class OffsetType {
        LeftHandTable,
        RightHandTable,
        Array,
    };
    struct Offset {
        OffsetType type;
        quint32 offset;
    };
    std::array<Offset, 3> offsets = {{
        { OffsetType::LeftHandTable, (quint32)leftHandTableOffset },
        { OffsetType::RightHandTable, (quint32)rightHandTableOffset },
        { OffsetType::Array, (quint32)arrayOffset },
    }};

    algo::sort_all_by_key(offsets, &Offset::offset);

    for (const auto offset : offsets) {
        if (offset.offset == 0) {
            continue;
        }

        parser.advanceTo(subtableStart + offset.offset);
        switch (offset.type) {
        case OffsetType::LeftHandTable: {
            parser.beginGroup("Left-hand Class Table");
            parser.read<GlyphId>("First glyph");
            const quint16 count = parser.read<UInt16>("Number of glyphs");
            parser.readBasicArray<UInt16>("Classes", count);
            parser.endGroup();
            break;
        }
        case OffsetType::RightHandTable: {
            parser.beginGroup("Right-hand Class Table");
            parser.read<GlyphId>("First glyph");
            const quint16 count = parser.read<UInt16>("Number of glyphs");
            parser.readBasicArray<UInt16>("Classes", count);
            parser.endGroup();
            break;
        }
        case OffsetType::Array: {
            parser.readBasicArray<Int16>("Kerning Values", rows * columns);
            break;
        }
        }
    }
}

static void parseFormat3(const quint32 subtableStart, const quint32 subtableSize, Parser &parser)
{
    // Apple fonts that are using this table: Skia.ttf

    const auto glyphCount = parser.read<UInt16>("Number of glyphs");
    const auto kernValues = parser.read<UInt8>("Number of kerning values");
    const auto leftHandClasses = parser.read<UInt8>("Number of left-hand classes");
    const auto rightHandClasses = parser.read<UInt8>("Number of right-hand classes");
    parser.read<UInt8>("Reserved");

    parser.readBasicArray<Int16>("Kerning Values", kernValues);
    parser.readBasicArray<UInt8>("Left-hand Classes", glyphCount);
    parser.readBasicArray<UInt8>("Right-hand Classes", glyphCount);
    parser.readBasicArray<UInt8>("Indices", int(leftHandClasses) * int(rightHandClasses));

    const auto left = int(subtableSize) - int(parser.offset() - subtableStart);
    parser.readPadding(left);
}

// https://docs.microsoft.com/en-us/typography/opentype/spec/kern
static void parseKernOpenType(Parser &parser)
{
    parser.read<UInt16>("Version");
    const auto numberOfTables = parser.read<UInt16>("Number of tables");
    parser.readArray("Subtables", numberOfTables, [&](const auto index){
        const auto subtableStart = parser.offset();

        parser.beginGroup(index);
        parser.read<UInt16>("Version");
        parser.read<UInt16>("Length");
        const auto format = parser.read<UInt8>("Format");
        parser.read<OpenTypeCoverage>("Coverage");

        switch (format) {
        case 0: parseFormat0(parser); break;
        case 2: parseFormat2(subtableStart, parser); break;
        default: throw QString("%1 is not a valid format").arg(format);
        }

        parser.endGroup("", QString("Format %1").arg(format));
    });
}

// https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6kern.html
static void parseKernApple(Parser &parser)
{
    parser.read<F16DOT16>("Version");
    const auto numberOfTables = parser.read<UInt32>("Number of tables");
    parser.readArray("Subtables", numberOfTables, [&](const auto index){
        const auto subtableStart = parser.offset();

        parser.beginGroup(index);
        const auto length = parser.read<UInt32>("Length");
        // Yes, the coverage and format order is inverted in AAT.
        parser.read<AppleCoverage>("Coverage");
        const auto format = parser.read<UInt8>("Format");
        parser.read<UInt16>("Tuple index");

        switch (format) {
        case 0: parseFormat0(parser); break;
        case 1: Format1::parse(length, parser); break;
        case 2: parseFormat2(subtableStart, parser); break;
        case 3: parseFormat3(subtableStart, length, parser); break;
        default: throw QString("%1 is not a valid format").arg(format);
        }

        parser.endGroup("", QString("Format %1").arg(format));
    });
}

void parseKern(Parser &parser)
{
    // The `kern` table has two variants: OpenType one and Apple one.
    // And they both have different headers.
    // There are no robust way to distinguish them, so we have to guess.
    //
    // The OpenType one has the first two bytes (UInt16) as a version set to 0.
    // While Apple one has the first four bytes (Fixed) set to 1.0
    // So the first two bytes in case of an OpenType format will be 0x0000
    // and 0x0001 in case of an Apple format.
    const auto version = parser.peek<UInt16>();
    if (version == 0) {
        parseKernOpenType(parser);
    } else {
        parseKernApple(parser);
    }
}
